libclamav/ole2_extract.c

```diff
@@ -586,6 +608,272 @@ static cl_error_t handler_enum(ole2_header_t *hdr, property_t *prop, const char
+}
+#pragma GCC diagnostic pop
+
+typedef struct {
+    bool velvet_sweatshop;
+
+    bool encrypted;
+
+    const char *encryption_type;
+
+} encryption_status_t;
+
+const char *const RC4_ENCRYPTION    = "RC4Encryption";
+const char *const XOR_OBFUSCATION   = "XORObfuscation";
+const char *const AES128_ENCRYPTION = "EncryptedWithAES128";
+const char *const AES192_ENCRYPTION = "EncryptedWithAES192";
+const char *const AES256_ENCRYPTION = "EncryptedWithAES256";
+
+const uint16_t XLS_XOR_OBFUSCATION    = 0;
+const uint16_t XLS_RC4_ENCRYPTION     = 1;
+const uint32_t MINISTREAM_CUTOFF_SIZE = 0x1000;
+
+static uint32_t get_stream_data_offset(ole2_header_t *hdr, const property_t *word_block, uint16_t sector)
+{
+    uint32_t offset      = (1 << hdr->log2_big_block_size);
+    uint32_t sector_size = offset;
+    uint32_t fib_offset  = 0;
+
+    if (word_block->size < MINISTREAM_CUTOFF_SIZE) {
+        fib_offset = offset + sector_size * hdr->sbat_root_start;
+        fib_offset += (word_block->start_block * (1 << hdr->log2_small_block_size));
+    } else {
+        fib_offset = offset + sector_size * sector;
+    }
+
+    return fib_offset;
+}
+
+/* See information about the File Information Block here
+ * https://learn.microsoft.com/en-us/openspecs/office_file_formats/ms-doc/26fb6c06-4e5c-4778-ab4e-edbf26a545bb
+ * for more information.
+ */
+static void test_for_encryption(const property_t *word_block, ole2_header_t *hdr, encryption_status_t *pEncryptionStatus)
+{
+
+    const uint8_t *ptr = NULL;
+    fib_base_t fib     = {0};
+
+    uint32_t fib_offset = get_stream_data_offset(hdr, word_block, word_block->start_block);
+
+    if ((size_t)(hdr->m_length) < (size_t)(fib_offset + sizeof(fib_base_t))) {
+        cli_dbgmsg("ERROR: Invalid offset for File Information Block %d (0x%x)\n", fib_offset, fib_offset);
+        return;
+    }
+
+    ptr = fmap_need_off_once(hdr->map, fib_offset, sizeof(fib_base_t));
+    if (NULL == ptr) {
+        cli_dbgmsg("ERROR: Invalid offset for File Information Block %d (0x%x)\n", fib_offset, fib_offset);
+        return;
+    }
+    copy_fib_base(&fib, ptr);
+
+#define FIB_BASE_IDENTIFIER 0xa5ec
+
+    if (FIB_BASE_IDENTIFIER != fib.wIdent) {
+        cli_dbgmsg("ERROR: Invalid identifier for File Information Block %d (0x%x)\n", fib.wIdent, fib.wIdent);
+        return;
+    }
+
+    /*TODO: Look into whether or not it's possible to determine the xor key when
+     * a document is obfuscated with xor
+     * (is_obfuscated function)
+     */
+    pEncryptionStatus->encrypted = is_encrypted(&fib);
+
+    if (is_obfuscated(&fib)) {
+        pEncryptionStatus->encryption_type = XOR_OBFUSCATION;
+    }
+}
+
+static bool read_uint16(const uint8_t *const ptr, uint32_t ptr_size, uint32_t *idx, uint16_t *dst)
+{
+    if (*idx + sizeof(uint16_t) >= ptr_size) {
+        return false;
+    }
+
+    memcpy(dst, &(ptr[*idx]), 2);
+    *dst = ole2_endian_convert_16(*dst);
+    *idx += sizeof(uint16_t);
+    return true;
+}
```

Comment:
As we discussed, I made the size_t change, but not the fmap change.  I really like the idea, but I feel it would be a lot of changes in this file, for something we aren't adding to fmap at this point.